README.md
# Circular

<p>
 <a href="https://flutter.dev/">
 <img src ="https://img.shields.io/badge/Flutter-02569B?style=for-the-badge&logo=flutter&logoColor=white">
</img>
 </a>
<a href="https://github.com/Mahmoudgarwallane/circular_slider">
 <img src ="https://img.shields.io/badge/GitHub-100000?style=for-the-badge&logo=github&logoColor=white">
</img>
</a>
  </p>

<img width=300 src ="https://user-images.githubusercontent.com/68555232/166114114-d7e47d13-3d8c-49ee-af62-c89da85764ec.gif">
</img>

## Features

- picking values in specific range
- displaying values in a a circular slider
## Getting started

#### add the package
```bash
flutter pub add circular
```
#### import the package

```bash
import 'package:circular/circular.dart';
```

## Usage

#### Circular slider

```dart
CircularSlider(
decoration: BoxDecoration(
    borderRadius: BorderRadius.circular(200),
    boxShadow: const [
      BoxShadow(
          offset: Offset(-10, -10),
          color: Colors.white,
          blurRadius: 20,
          spreadRadius: 1),
      BoxShadow(
          offset: Offset(10, 10),
          color: Color.fromARGB(255, 158, 158, 158),
          blurRadius: 20,
          spreadRadius: 1)
    ]),
maxValue: 100,
radius:100, 
child: Text(
  value.toString(),
  style: TextStyle(fontSize: 30),
),
color: Color(0xffEEEEEE),
sliderColor: Color(0xff62CBDA),
unSelectedColor: Color(0xffD7DEE7),
onDrag: (value) {
  setState(() {
    this.value = value;
  });
},
),
```
#### Circular Viewer
```dart
CircularViewer(
  decoration: BoxDecoration(
      borderRadius: BorderRadius.circular(200),
      boxShadow: const [
        BoxShadow(
            offset: Offset(-10, -10),
            color: Colors.white,
            blurRadius: 20,
            spreadRadius: 1),
        BoxShadow(
            offset: Offset(10, 10),
            color: Color.fromARGB(255, 158, 158, 158),
            blurRadius: 20,
            spreadRadius: 1)
      ]),
  value: 50,
  maxValue: 100,
  radius: 100,
  textStyle: TextStyle(fontSize: 30),
  color: Color(0xffEEEEEE),
  sliderColor: Color(0xff62CBDA),
  unSelectedColor: Color(0xffD7DEE7),
)
```

check out the [full example](https://github.com/Mahmoudgarwallane/circular_slider/blob/master/example/example.dart)

## Additional information
### story 
after I made the [the magic circle ](https://github.com/Mahmoudgarwallane/Magic-circle) I decided to make it useful so I made this package 

### Contributing
Contributions are always welcome!
### License
[MIT LICENSE](LICENSE)


CHANGELOG.md
## 0.0.2

* Cicular Slider and Viewer

## 0.0.3
* updating package name

## 0.0.4

* adding platforms support

## 0.0.5
* fixing repetitive animation

lib/src/magic_circle.dart
import 'package:flutter/material.dart';
import 'dart:math';

class Circle extends CustomPainter {
  final Color color;
  final Color sliderColor;
  final Color unSelectedColor;
  final double angle;
  Circle(
      {required this.angle,
      required this.color,
      required this.sliderColor,
      required this.unSelectedColor});
  @override
  void paint(Canvas canvas, Size size) {
    var center = const Offset(0, 0);
    canvas.translate(size.width / 2, size.height / 2);
    var radius = size.width / 2;
    var circleBrush = Paint()..color = color;
    var unSelectedAreaBrush = Paint()..color = unSelectedColor;
    var sliderBrush = Paint()
      ..color = sliderColor
      ..style = PaintingStyle.stroke
      ..strokeWidth = 25
      ..strokeCap = StrokeCap.round;
    canvas.drawCircle(center, radius, circleBrush);
    canvas.drawCircle(center, radius - 20, unSelectedAreaBrush);
    canvas.drawArc(
        Rect.fromCircle(center: const Offset(0, 0), radius: radius - 32),
        0, //radians
        -angle * pi / 180, //radians
        false,
        sliderBrush);
    canvas.drawCircle(center, radius - 45, circleBrush);
  }

  @override
  bool shouldRepaint(covariant CustomPainter oldDelegate) {
    return true;
  }
}


lib/src/circular_slider.dart
part of circular_slider;

class CircularSlider extends StatefulWidget {
  final Function(int value) onDrag;
  final Color color;
  final Widget child;
  final Color sliderColor;
  final Color unSelectedColor;
  final double radius;
  final double maxValue;
  final BoxDecoration? decoration;
  //! parent size must be equal or higher than radius * 2

  const CircularSlider(
      {Key? key,
      required this.maxValue,
      this.decoration,
      required this.radius,
      required this.onDrag,
      required this.child,
      required this.color,
      required this.sliderColor,
      required this.unSelectedColor})
      : super(key: key);

  @override
  State<CircularSlider> createState() => _CircularSliderState();
}

class _CircularSliderState extends State<CircularSlider> {
  double angle = 0;
  void onPanUpdate(DragUpdateDetails details) {
    Offset coordinates = details.localPosition;

    setState(() {
      var center = (widget.radius * 2) / 2;
      Offset pureCoordinates = Offset(((coordinates.dx - center) / center),
          ((coordinates.dy - center) / center) * -1);
      var angleTan = (pureCoordinates.dy.abs()) / (pureCoordinates.dx.abs());

      (pureCoordinates.toString());
      double minValue =
          sqrt((pow(pureCoordinates.dx, 2) + pow(pureCoordinates.dy, 2)));

      if ((minValue > 0.6)) {
        angle = atan(angleTan) * 180 / pi;
        angle = angle.roundToDouble();
        if (pureCoordinates.dx.isNegative && !pureCoordinates.dy.isNegative) {
          angle = 90 - angle;
          angle += 90;
        }
        if (pureCoordinates.dx.isNegative && pureCoordinates.dy.isNegative) {
          angle += 180;
        }
        if (!pureCoordinates.dx.isNegative && pureCoordinates.dy.isNegative) {
          angle = 90 - angle;
          angle += 270;
        }
      }

      widget.onDrag(
        angle * widget.maxValue ~/ 360,
      );
    });
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: Alignment.center,
      children: [
        Container(
          width: widget.radius * 2,
          height: widget.radius * 2,
          decoration: widget.decoration ,
          child: GestureDetector(
            onPanUpdate: onPanUpdate,
            child: CustomPaint(
              painter: Circle(
                  angle: angle,
                  color: widget.color,
                  unSelectedColor: widget.unSelectedColor,
                  sliderColor: widget.sliderColor),
            ),
          ),
        ),
        widget.child
      ],
    );
  }
}


lib/src/circular_viewer.dart
part of circular_slider;

class CircularViewer extends StatefulWidget {
  final Color color;
  final Color sliderColor;
  final Color unSelectedColor;
  final double value;
  final double radius;
  final double maxValue;
  final TextStyle textStyle;
  final BoxDecoration? decoration;
  const CircularViewer(
      {Key? key,
      required this.maxValue,
      this.textStyle = const TextStyle(fontSize: 20),
      this.decoration,
      required this.radius,
      required this.color,
      required this.value,
      required this.sliderColor,
      required this.unSelectedColor})
      : super(key: key);

  @override
  State<CircularViewer> createState() => _CircularViewerState();
}

class _CircularViewerState extends State<CircularViewer>
    with SingleTickerProviderStateMixin {
  late AnimationController controller;
  int animatedValue = 0;
  late Animation<int> animation;
  @override
  void initState() {
    super.initState();

    controller =
        AnimationController(vsync: this, duration: const Duration(seconds: 2));
    animation = IntTween(begin: 0, end: widget.value.toInt()).animate(
        CurvedAnimation(parent: controller, curve: Curves.fastOutSlowIn));
    controller.forward();
    animation.addListener(() {
      setState(() {});
    });
  }

  @override
  void dispose() {
    controller.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return Stack(
      alignment: Alignment.center,
      children: [
        Container(
          decoration: widget.decoration,
          width: widget.radius * 2,
          height: widget.radius * 2,
          child: CustomPaint(
            painter: Circle(
                angle: animation.value.toDouble() * 360 / widget.maxValue,
                color: widget.color,
                unSelectedColor: widget.unSelectedColor,
                sliderColor: widget.sliderColor),
          ),
        ),
        Text(
          animation.value.toString(),
          style: widget.textStyle,
        )
      ],
    );
  }
}


example/example.dart
import 'package:flutter/material.dart';
import 'package:circular/circular.dart';

void main() {
  runApp(const MaterialApp(
    debugShowCheckedModeBanner: false,
    home: Scaffold(body: Center(child: Home())),
  ));
}
 
class Home extends StatefulWidget {
  const Home({Key? key}) : super(key: key);

  @override
  State<Home> createState() => _HomeState();
}

class _HomeState extends State<Home> {
  int angle = 0;
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: SafeArea(
        child: Container(
          color: const Color(0xffEEEEEE),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceAround,
            children: [
              SizedBox(
                width: 200, //! this should not be less than
                height: 200,
                child: CircularSlider(
                  decoration: BoxDecoration(
                      borderRadius: BorderRadius.circular(200),
                      boxShadow: const [
                        BoxShadow(
                            offset: Offset(-10, -10),
                            color: Colors.white,
                            blurRadius: 20,
                            spreadRadius: 1),
                        BoxShadow(
                            offset: Offset(10, 10),
                            color: Color.fromARGB(255, 158, 158, 158),
                            blurRadius: 20,
                            spreadRadius: 1)
                      ]),
                  maxValue: 100,
                  radius:
                      100, //! this one multiplied by 2  ( sized box width and height >= radius * 2)
                  child: Text(
                    angle.toString(),
                    style: const TextStyle(fontSize: 30),
                  ),
                  color: const Color(0xffEEEEEE),
                  sliderColor: const Color(0xff62CBDA),
                  unSelectedColor: const Color(0xffD7DEE7),
                  onDrag: (angle) {
                    setState(() {
                      this.angle = angle;
                    });
                  },
                ),
              ),
              CircularViewer(
                decoration: BoxDecoration(
                    borderRadius: BorderRadius.circular(200),
                    boxShadow: const [
                      BoxShadow(
                          offset: Offset(-10, -10),
                          color: Colors.white,
                          blurRadius: 20,
                          spreadRadius: 1),
                      BoxShadow(
                          offset: Offset(10, 10),
                          color: Color.fromARGB(255, 158, 158, 158),
                          blurRadius: 20,
                          spreadRadius: 1)
                    ]),
                value: 50,
                maxValue: 100,
                radius: 100,
                textStyle: const TextStyle(fontSize: 30),
                color: const Color(0xffEEEEEE),
                sliderColor: const Color(0xff62CBDA),
                unSelectedColor: const Color(0xffD7DEE7),
              )
            ],
          ),
          alignment: Alignment.center,
        ),
      ),
    );
  }
}
